api/src/main/java/io/minio/ComposeSource.java

```diff
@@ -0,0 +1,189 @@
+
+  /**
+   * Create new ComposeSource for given bucket, object, offset, length, headerMap and CopyConditions.
+   */
+  public ComposeSource(String bucketName, String objectName, Long offset, Long length,
+      Map<String, String> headerMap, CopyConditions copyConditions) throws InvalidArgumentException {
+    this(bucketName, objectName, offset, length, headerMap, copyConditions, null);
+  }
+
+  /**
+   * Creates new ComposeSource for given bucket, object, offset, length, headerMap, CopyConditions
+   * and server side encryption.
+   *
+   * @throws InvalidArgumentException    upon invalid value is passed to a method.
+   */
+  public ComposeSource(String bucketName, String objectName, Long offset, Long length, Map<String, String> headerMap,
+      CopyConditions copyConditions, ServerSideEncryption sse) throws InvalidArgumentException {
+    if (bucketName == null) {
+      throw new InvalidArgumentException("Source bucket name cannot be empty");
+    }
+
+    if (objectName == null) {
+      throw new InvalidArgumentException("Source object name cannot be empty");
+    }
+
+    if (offset != null && offset < 0) {
+      throw new InvalidArgumentException("Offset cannot be negative");
+    }
+
+    if (length != null && length < 0) {
+      throw new InvalidArgumentException("Length cannot be negative");
+    }
+
+    if (length != null && offset == null) {
+      offset = 0L;//Long.valueOf(0);//new Long(0);
+    }
+
+    this.bucketName = bucketName;
+    this.objectName = objectName;
+    this.offset = offset;
+    this.length = length;
+    this.headerMap = headerMap;
+    this.copyConditions = copyConditions;
+    this.sse = sse;
+  }
+
+  /**
+   * Constructs header  .
+   *
+   */
+  public void buildHeaders(long objectSize, String etag) throws InvalidArgumentException {
+    if (offset != null && offset >= objectSize) {
+      throw new InvalidArgumentException("source " + bucketName + "/" + objectName + ": offset " + offset
+        + " is beyond object size " + objectSize);
+    }
+
+    if (length != null) {
+      if (length >= objectSize) {
+        throw new InvalidArgumentException("source " + bucketName + "/" + objectName + ": length " + length
+          + " is beyond object size " + objectSize);
+      }
+
+      if (offset + length > objectSize) {
+        throw new InvalidArgumentException("source " + bucketName + "/" + objectName + ": compose size "
+          + (offset + length) + " is beyond object size " + objectSize);
+      }
+    }
+
+    Map<String, String> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+    headers.put("x-amz-copy-source", S3Escaper.encodePath(bucketName + "/" + objectName));
+    headers.put("x-amz-copy-source-if-match", etag);
+
+    if (headerMap != null) {
+      headers.putAll(headerMap);
+    }
+
+    if (copyConditions != null) {
+      headers.putAll(copyConditions.getConditions());
+    }
+
+    if (sse != null) {
+//      sse.marshal(headers);
+      headers = sse.headers();
```

Comment:
1. remove unnecessary comments
2. you need to use `sse.copySourceHeaders()`

Reply:
done.